examples/custom_entities.rs

```diff
@@ -1,76 +1,207 @@
-//! This example demonstrate how custom entities can be extracted from the DOCTYPE!,
-//! and later use to decode text and attribute values.
+//! This example demonstrate how custom entities can be extracted from the DOCTYPE,
+//! and later use to:
+//! - insert new pieces of document (particular case - insert only textual content)
+//! - decode attribute values
 //!
 //! NB: this example is deliberately kept simple:
 //! * it assumes that the XML file is UTF-8 encoded (custom_entities must only contain UTF-8 data)
 //! * it only handles internal entities;
 //! * the regex in this example is simple but brittle;
 //! * it does not support the use of entities in entity declaration.
 
-use std::collections::HashMap;
+use std::borrow::Cow;
+use std::collections::{HashMap, VecDeque};
+use std::str::from_utf8;
 
-use quick_xml::escape::resolve_predefined_entity;
-use quick_xml::events::Event;
+use quick_xml::encoding::Decoder;
+use quick_xml::errors::Error;
+use quick_xml::escape::EscapeError;
+use quick_xml::events::{BytesEnd, BytesStart, BytesText, Event};
+use quick_xml::name::QName;
 use quick_xml::reader::Reader;
 use regex::bytes::Regex;
 
-const DATA: &str = r#"
+use pretty_assertions::assert_eq;
 
-    <?xml version="1.0"?>
-    <!DOCTYPE test [
-    <!ENTITY msg "hello world" >
-    ]>
-    <test label="&msg;">&msg;</test>
+struct MyReader<'i> {
```

Comment:
We very much need to make this functionality usable without the user needing to write or copy over a custom wrapper, IMO.  I believe that was the plan, I just want to make sure.

Reply:
My point just being that the current example as-is is a bit much to expect people to implement or copy and paste, and I just want to check my understanding that it's not a long-term solution.

Reply:
I have not really decided yet whether to leave this example as it is to demonstrate how the reader stack can be implemented if for some reason the standard solution does not work, or rewrite it to a new API. If it remains as it is, then there will be a mention of the standard way